altair/utils/core.py

```diff
@@ -527,6 +527,22 @@ def parse_shorthand(
     if isinstance(data, pd.DataFrame) and "type" not in attrs:
         if "field" in attrs and attrs["field"] in data.columns:
             attrs["type"] = infer_vegalite_type(data[attrs["field"]])
+
+    # If an unescaped colon is still present, it's often due to an incorrect data type specification
+    # but could also be due to using a column name with ":" in it.
+    if (
+        "field" in attrs
+        and ":" in attrs["field"]
+        and attrs["field"][attrs["field"].rfind(":") - 1] != "\\"
+    ):
+        raise ValueError(
+            '"{}" '.format(attrs["field"].split(":")[-1])
+            + "is not one of the valid encoding data types: {}.".format(
+                ", ".join(TYPECODE_MAP.values())
+            )
+            + "\nFor more details, see https://altair-viz.github.io/altair-docs/user_guide/encodings/index.html#encoding-data-types. "
+            + "(If you are trying to use a column name that contains a colon, prefix it with a backslash, \\:)"
+        )
```

Comment:
I think it does help improve the error message since there will now be a clear error message when the type is incorrect. Otherwise, this would be a potential issue to mention in the other error message as well which makes it more vague. If you prefer I can break this out into its own PR to discuss more?

Reply:
Regarding escaping the colon, I don't really like the breaking change, but I also agree with you that the current behaviour can lead to confusing situations. 

Can you include an example in the `ValueError` how this is supposed to look like with a real example? Maybe for a column that is named as such `"release_date:theater"`, how is this supposed to look like when escaped? Would this change also be required if a person does not use the shorthand `:T`, but defines it as `"release_date:theater", type="temporal"` instead?

Reply:
Yeah, I agree that a breaking change is not ideal, but at the same time we are planning a bump in the major version number with the 5.0 release and that is an indication that there might be breaking changes. I updated the error message to include a column name and it now reads:

> If you are trying to use a column name that contains a colon, prefix it with a backslash; for example "column\:name" instead of "column:name".

---

Essentially what this PR does is treating `:` the same as other special characters such as `.`, `[`, `-`, etc. The one difference being that `:` is a special character at the altair level, whereas the others are special characters at the VL level. For all these special characters, we need to both escape them and specify the type (although that could be improved on in a later PR). Here are three examples:

## `:`

```py
import pandas as pd
import altair as alt

source = pd.DataFrame({
    'release_year:national': [2000, 2001, 2004],
    'release_year:international': [2001, 2001, 2005],
    'color-': [2001, 2001, 2005]
})

alt.Chart(source).mark_point().encode(
    x='release_year\:national:Q',
    y=alt.Y('release_year\:international', type='quantitative'),
)
```

## `[`

```py
import altair as alt
import numpy as np
import pandas as pd

x = np.arange(100)

xname = "x [mV"
xname_escaped = "x \[mV"

source = pd.DataFrame({xname: x})

alt.Chart(source).mark_line().encode(
    x=alt.X(field=xname_escaped, type="quantitative"),
    y=xname_escaped + ':Q',
)
```

## `.`

```py
import altair as alt
import pandas as pd
import numpy as np

source = pd.DataFrame({"x.x": np.arange(10)})

alt.Chart(source).mark_line().encode(
    x="x\.x:Q",
    y=alt.Y("x\.x", type='quantitative')
)
```

And some old issues discussing special characters (posting as a reference for myself later too):

- https://github.com/altair-viz/altair/issues/284
- https://github.com/altair-viz/altair/issues/2494
- https://github.com/altair-viz/altair/issues/2600